File: projects/engineering.purdue.edu/kak/scriptingwithobjects/swocode/chap14/MultiCustomerAccount.py
#!/usr/bin/env python

###  MultiCustomerAccount.py

import threading    
import time

import random                  
ran = random.Random()          

lock = threading.RLock()                                             #(A)
cv = threading.Condition(lock)                                       #(B)

#----------------------   utility class Account   ----------------------
class Account( object ):                                             #(C)

    def __init__( self ):                                            #(D)
        self.balance = 0                                             #(E)

    def deposit( self, dep ):                                        #(F)
        lock.acquire()                                               #(G)
        self.balance += dep                                          #(H)
        cv.notifyAll()                                               #(I)
        lock.release()                                               #(J)

    def withdraw( self, draw ):                                      #(K)
        lock.acquire()                                               #(L)
        while ( self.balance < draw ):                               #(M)
            cv.wait()                                                #(N)
        self.balance -= draw                                         #(O)
        lock.release()                                               #(P)

#----------------  multithreadable class Depositor   ------------------- 
class Depositor( threading.Thread ):                                 #(Q)

    def __init__( self, acnt ):                                      #(R)
        self.account = acnt                                          #(S)
        threading.Thread.__init__(self)                              #(T)

    def run(self):                                                   #(U)
        i = 0                                                        #(V)
        while 1:                                                     #(W)
            x = ran.randint(0,10)                                    #(X)
            self.account.deposit( x )                                #(Y)
            i += 1                                                   #(Z)
            if i % 10 == 0:                                          #(a)
                print "balance after deposits %d" % \
                                         self.account.balance        #(b)
            time.sleep( 1 )                                          #(c)

#----------------  multithreadable class Withdrawer   ------------------ 
class Withdrawer( threading.Thread ):                                #(d)

    def __init__( self, acnt ):                                      #(e)
        self.account = acnt                                          #(f)
        threading.Thread.__init__(self)                              #(g)

    def run(self):                                                   #(h)
        i = 0                                                        #(i)
        while 1:                                                     #(j)
            x = ran.randint(0,10)                                    #(k)
            self.account.withdraw( x )                               #(l)
            i += 1                                                   #(m)
            if i % 10 == 0:                                          #(n)
                print "balance after withdrawals %d" % \
                                        self.account.balance         #(o)
            time.sleep( 1 )                                          #(p)

#--------------------------   launch threads   ------------------------- 

acct = Account()                                                     #(q)

depositors = [ Depositor(acct) for i in range(3) ]                   #(r)
withdrawers = [ Withdrawer(acct) for i in range(3) ]                 #(s)

map( lambda x: x.start(), depositors + withdrawers )                 #(t)

File: projects/demos/engineering.purdue.edu/scriptingwithobjects/swocode/chap14/MultiCustomerAccount.pl
#!/usr/bin/perl -w

###  MultiCustomerAccount.pl

use strict;
use threads;                    
use threads::shared;            

my $account_balance : shared = 0;                                    #(A)
my $cv : shared;                                                     #(B)

foreach (1 .. 3) {                                                   #(C)
    threads->new( \&multiple_deposits );                             #(D)
}
foreach (1 .. 3) {                                                   #(E)
    threads->new( \&multiple_withdrawals );                          #(F)
}

foreach (threads->list) { $_->join }                                 #(G)

sub deposit {                                                        #(H)
    my $dep = shift;                                                 #(I)
    lock( $cv );                                                     #(J)
    $account_balance += $dep;                                        #(K)
    cond_signal( $cv );                                              #(L)
}

sub withdraw {                                                       #(M)
    my $draw = shift;                                                #(N)
    lock( $cv );                                                     #(O)
    while ( $account_balance < $draw ) {                             #(P)
        cond_wait( $cv );                                            #(Q)
    }                            
    $account_balance -= $draw;                                       #(R)
}

sub multiple_deposits {                                              #(S)
    my $i = 0;                                                       #(T)
    my $x;                                                           #(U)
    while ( 1 ) {                                                    #(V)
        $x = rand( 10 );                                             #(W)
        deposit( $x );                                               #(X)
        if ( $i++ % 10 == 0 ) {                                      #(Y)
            my $threadid = threads->self->tid;                       #(Z)
            printf( 
              "balance after $i deposits by thread $threadid: \t\t%d\n", 
                               $account_balance );                   #(a)
        }
        keepBusy( 1 );                                               #(b)
    }
}

sub multiple_withdrawals {                                           #(c)
    my $i = 0;                                                       #(d)
    my $x;                                                           #(e)
    while ( 1 ) {                                                    #(f)
        $x = rand( 10 );                                             #(g)
        withdraw( $x );                                              #(h)
        if ( $i++ % 10 == 0 )  {                                     #(i)
            my $threadid = threads->self->tid;                       #(j)
            printf( 
            "balance after $i withdrawals by thread $threadid: \t%d\n", 
                                   $account_balance );               #(k)
        }
        keepBusy( 1 );                                               #(l)
    }
}

sub keepBusy {                                                       #(m)
    my $how_many_seconds = shift;                                    #(n)
    my $curr = time;                                                 #(o)
    while ( time < $curr + $how_many_seconds ) {}                    #(p)
}
